tracer/src/Datadog.Trace/AppSec/Waf/Context.cs

```diff
@@ -16,19 +16,18 @@ namespace Datadog.Trace.AppSec.Waf
     internal class Context : IContext
     {
         private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<Context>();
-        private readonly IntPtr contextHandle;
-        private readonly WafNative wafNative;
-        private readonly Encoder encoder;
+        private readonly IntPtr _contextHandle;
+        private readonly Waf _waf;
         private readonly List<Obj> argCache = new();
         private readonly Stopwatch _stopwatch;
+        private object lockObj = new();
```

Comment:
I think it's a good idea to document which variables are protected by this lock.

It seems we only protect the `_waf` but I think we should protect `_contextHandle` & `argCache` since these both contain unmanaged objects that shouldn't be accessed currently.

Two further things occur to me about the locking of `_waf`:
* I wonder if it's enough to lock access to the `_waf`, this reference is shared with `Security.cs` and I would have thought the creation and destruction operations there also need locking.
* since the WAF can support concurrent access, it's the creation / destruction operations that need protecting, perhaps we should be using `readerwriterlock` which is optimized for multiple readers: https://learn.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlock?view=net-7.0
I also

Reply:
that was actually to protect the _context handle access on the run and on destroy.  I changed the lock to also include the encoding on `_argCache` in d358405a702f8b3b907075d68c28bb3146c7a794 and renamed it.

Reply:
I love the idea of `readerwriterlock ` , seems to be _exactly_ what we need if it's performant enough. I didnt know so much about it! Maybe we could merge this PR for now, as it has quite some refactoring, and later, have a PR for this strategy specifically that would replace/complete the protection we have here? (right now we have nothing ^^)

Reply:
I think we should use `ReaderWriterLockSlim` instead of `ReaderWriterLock` if we go that way ([as per the recommendation](https://learn.microsoft.com/en-us/dotnet/api/system.threading.readerwriterlockslim?view=net-7.0#remarks)), but  as also noted there, it's not thread-abort safe, so not safe to use in .NET Framework

Reply:
good point! done [here](https://github.com/DataDog/dd-trace-dotnet/tree/anna/asm/lock-context/tracer/src/Datadog.Trace/AppSec/Concurrency), we use slim for core and normal for .net framework